pkg/cloud/aws/services/ec2/instances_test.go

```diff
@@ -553,3 +555,472 @@ vuO9LYxDXLVY9F7W4ccyCqe27Cj1xyAvdZxwhITrib8Wg5CMqoRpqTw5V3+TpA==
+	type args struct {
+		initConfig kubeadmv1beta1.InitConfiguration
+		machine    *clusterv1.Machine
+	}
+	tests := []struct {
+		name     string
+		args     args
+		expected kubeadmv1beta1.InitConfiguration
+	}{
+		{
+			name: "simple",
+			args: args{
+				initConfig: kubeadmv1beta1.InitConfiguration{},
+				machine:    &clusterv1.Machine{},
+			},
+			expected: kubeadmv1beta1.InitConfiguration{
+				NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+					Name:      "{{ ds.meta_data.hostname }}",
+					CRISocket: "/var/run/containerd/containerd.sock",
+					KubeletExtraArgs: map[string]string{
+						"cloud-provider": "aws",
+					},
+				},
+			},
+		},
+		{
+			name: "with node taints",
+			args: args{
+				initConfig: kubeadmv1beta1.InitConfiguration{},
+				machine: &clusterv1.Machine{
+					Spec: clusterv1.MachineSpec{
+						Taints: []corev1.Taint{{
+							Key:    "my-property",
+							Value:  "special",
+							Effect: corev1.TaintEffectNoSchedule,
+						}},
+					},
+				},
+			},
+			expected: kubeadmv1beta1.InitConfiguration{
+				NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+					Name:      "{{ ds.meta_data.hostname }}",
+					CRISocket: "/var/run/containerd/containerd.sock",
+					KubeletExtraArgs: map[string]string{
+						"cloud-provider": "aws",
+					},
+					Taints: []corev1.Taint{{
+						Key:    "my-property",
+						Value:  "special",
+						Effect: corev1.TaintEffectNoSchedule,
+					}},
+				},
+			},
+		},
+		{
+			name: "with cni options",
+			args: args{
+				initConfig: kubeadmv1beta1.InitConfiguration{
+					NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+						KubeletExtraArgs: map[string]string{
+							"cni-bin-dir":  "/opt/cni/bin",
+							"cni-conf-dir": "/etc/cni/net.d",
+						},
+					},
+				},
+				machine: &clusterv1.Machine{},
+			},
+			expected: kubeadmv1beta1.InitConfiguration{
+				NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+					Name:      "{{ ds.meta_data.hostname }}",
+					CRISocket: "/var/run/containerd/containerd.sock",
+					KubeletExtraArgs: map[string]string{
+						"cloud-provider": "aws",
+						"cni-bin-dir":    "/opt/cni/bin",
+						"cni-conf-dir":   "/etc/cni/net.d",
+					},
+				},
+			},
+		},
+		{
+			name: "with cri socket",
+			args: args{
+				initConfig: kubeadmv1beta1.InitConfiguration{
+					NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+						CRISocket: "/var/run/my-cri.sock",
+					},
+				},
+				machine: &clusterv1.Machine{},
+			},
+			expected: kubeadmv1beta1.InitConfiguration{
+				NodeRegistration: kubeadmv1beta1.NodeRegistrationOptions{
+					Name:      "{{ ds.meta_data.hostname }}",
+					CRISocket: "/var/run/my-cri.sock",
+					KubeletExtraArgs: map[string]string{
+						"cloud-provider": "aws",
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			setInitConfigurationOptions(&tt.args.initConfig, tt.args.machine)
```

Comment:
There is no logic in `SetInitConfigurationOptions` so I'm a little confused as to what this is testing. Because we've tested the builder we can assume that this function works, but what do you think?

The only function you've added where there is logic is `stringOrDefault` which we can see is both easily testable (but doesn't necessarily need tests) and is unlikely to contain a bug.

Reply:
We had tested all of the individual pieces, but we hadn't tested their composition anywhere. The logic of "an Init configuration must include these pieces, but these pieces should be user overridable" that gets made up by which options end up getting passed with what arguments into the builder.

If we'd had a test like this "before" (a foolhardy assertion, I know 😆) I think it would've caught #823 .

Reply:
ahhh ok got it 👍 totally approve! thank you for explaining!